Address Safer CPP Warnings in AnimationEffect.cpp

stwrt opened the pull request:
#### 04976445ff23b60636e104853a41d9225a5e3839
<pre>
Address Safer CPP Warnings in AnimationEffect.cpp
<a href="https://bugs.webkit.org/show_bug.cgi?id=290553">https://bugs.webkit.org/show_bug.cgi?id=290553</a>
<a href="https://rdar.apple.com/problem/148041002">rdar://problem/148041002</a>

Reviewed by Ryosuke Niwa.

Address Safer CPP Warnings in AnimationEffect.cpp

* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/animation/AnimationEffect.cpp:
(WebCore::AnimationEffect::getBindingsTiming const):
(WebCore::AnimationEffect::getBindingsComputedTiming):
(WebCore::AnimationEffect::getComputedTiming):
(WebCore::AnimationEffect::bindingsUpdateTiming):
(WebCore::AnimationEffect::updateTiming):
(WebCore::AnimationEffect::updateComputedTimingPropertiesIfNeeded):

Canonical link: <a href="https://commits.webkit.org/292823@main">https://commits.webkit.org/292823@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/60fa18f93487f6c25de577bab511f34018a04039

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97228 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/16851 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/7062 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/102309 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/47753 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/99273 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/17144 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/25301 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/74075 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31273 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100231 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/12961 "Found 1 new test failure: fast/forms/ios/focus-input-via-button.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/87926 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54420 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/12714 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/5805 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/47195 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/82766 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/5883 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/104331 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/24304 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/17728 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/83118 "Passed tests") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/24679 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/84058 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82532 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20757 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/4758 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/17813 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/24267 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/29418 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/24090 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/27402 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/25663 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

graouts commented on Source/WebCore/animation/AnimationEffect.cpp:
```diff
@@ -73,7 +73,7 @@ EffectTiming AnimationEffect::getBindingsTiming() const
         timing.duration = autoAtom();
     timing.direction = m_timing.direction;
-    timing.easing = m_timing.timingFunction->cssText();
+    timing.easing = m_timing.timingFunction.get()->cssText();
```
`m_timing.timingFunction` is a `RefPtr<TimingFunction>`. Why is this not safe to call the `TimingFunction::cssText()` method on and how calling it on the raw pointer safer?

rniwa replied:
Yeah, I don't think this is correct. We need to store it in a local stack variable which is RefPtr instead.
The problem with calling a non-trivial function like `cssText` is that it could alter memory state. Out of caution, we need to store it in RefPtr in stack while making a function call.
See https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines#when-calling-a-non-trivial-member-function-on-an-object-hold-a-smart-pointer-to-the-object-on-the-stack

rniwa commented on Source/WebCore/animation/AnimationEffect.cpp:
```diff
@@ -134,7 +134,7 @@ ComputedEffectTiming AnimationEffect::getComputedTiming(std::optional<WebAnimati
     computedTiming.duration = computedDuration;
     computedTiming.direction = m_timing.direction;
-    computedTiming.easing = m_timing.timingFunction->cssText();
+    computedTiming.easing = m_timing.timingFunction.get()->cssText();
```
Ditto. You need to store this value in a stack allocated RefPtr before calling cssText.

rniwa commented on Source/WebCore/animation/AnimationEffect.cpp:
```diff
@@ -249,7 +249,7 @@ ExceptionOr<void> AnimationEffect::updateTiming(Document& document, std::optiona
 
     if (m_animation)
-        m_animation->effectTimingDidChange();
+        m_animation.get()->effectTimingDidChange();
```
Need store m_animation in a RefPtr in stack.

rniwa commented on Source/WebCore/animation/AnimationEffect.cpp:
```diff
@@ -449,7 +449,7 @@ void AnimationEffect::updateComputedTimingPropertiesIfNeeded()
 
         if (RefPtr scrollTimeline = dynamicDowncast<ScrollTimeline>(timeline)) {
-            auto interval = scrollTimeline->intervalForAttachmentRange(m_animation->range());
+            auto interval = scrollTimeline->intervalForAttachmentRange(m_animation.get()->range());
```
Ditto.

rniwa reviewed (changes requested):
Looks like we're missing fixes for line 62, 104, and 153. Once you fixed those, you should be able to remove the file from the expectation files under `Source/WebCore/SaferCPPExpectations/`.

rniwa commented on Source/WebCore/animation/AnimationEffect.cpp:
```diff
@@ -73,7 +74,7 @@ EffectTiming AnimationEffect::getBindingsTiming() const
         timing.duration = autoAtom();
     timing.direction = m_timing.direction;
-    timing.easing = m_timing.timingFunction->cssText();
+    timing.easing = timingFunction->cssText();
```
I'd just write this as this instead:
```cpp
timing.easing = RefPtr { m_timing.timingFunction }->cssText();
```

rniwa commented on Source/WebCore/animation/AnimationEffect.cpp:
```diff
@@ -127,14 +128,15 @@ ComputedEffectTiming AnimationEffect::getComputedTiming(std::optional<WebAnimati
     computedTiming.duration = computedDuration;
     computedTiming.direction = m_timing.direction;
-    computedTiming.easing = m_timing.timingFunction->cssText();
+    computedTiming.easing = timingFunction->cssText();
```
Ditto.

rniwa commented on Source/WebCore/animation/AnimationEffect.cpp:
```diff
@@ -248,8 +250,9 @@ ExceptionOr<void> AnimationEffect::updateTiming(Document& document, std::optiona
-    if (m_animation)
-        m_animation->effectTimingDidChange();
+    RefPtr animation = m_animation.get();
+    if (animation)
```
animation should be declared within if like so:
```cpp
if (RefPtr animation = m_animation)
```

rniwa commented on Source/WebCore/animation/AnimationEffect.cpp:
```diff
@@ -248,8 +248,8 @@ ExceptionOr<void> AnimationEffect::updateTiming(Document& document, std::optiona
 
-    if (m_animation)
-        m_animation->effectTimingDidChange();
+    if (RefPtr animation = m_animation)
```
oh, you need to call .get() here.

stwrt committed:
Address Safer CPP Warnings in AnimationEffect.cpp
https://bugs.webkit.org/show_bug.cgi?id=290553
rdar://problem/148041002

Reviewed by Ryosuke Niwa.

Address Safer CPP Warnings in AnimationEffect.cpp

* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/animation/AnimationEffect.cpp:
(WebCore::AnimationEffect::getBindingsTiming const):
(WebCore::AnimationEffect::getBindingsComputedTiming):
(WebCore::AnimationEffect::getComputedTiming):
(WebCore::AnimationEffect::bindingsUpdateTiming):
(WebCore::AnimationEffect::updateTiming):
(WebCore::AnimationEffect::updateComputedTimingPropertiesIfNeeded):

Canonical link: https://commits.webkit.org/292823@main
